### ISSUE TITLE
TreeView reveal creates a race condition with `TreeDataProvider#getChildren`

### ISSUE THREAD

**Issue by EhabY** [microsoft/vscode#192055]
If the tree provider is loading the tree and `TreeView#reveal` is revealing an element that is not loaded yet (so the parents are loaded) then a race condition can cause the view to break with the message "Element with id X is already registered".

I have also tried to create a promise queue as suggested in [this comment](https://github.com/microsoft/vscode/issues/142328#issuecomment-1046907141) but the issue still exists. I think this issue is caused by `reveal` and `refresh` of the tree, like it's not reveal statements causing it but rather the tree is not synchronized with reveal.

See the following example repo: https://github.com/EhabY/TreeViewRevealExample/tree/main
1. Run the extension
2. Quickly run the command `Test View: Reveal` twice in a row

https://github.com/microsoft/vscode/assets/12927950/d68f2f7a-49a0-4867-ac93-ba168306a014

**Comment by EhabY** [comment]
Pinging @alexr00 so it does not get lost 🙏

**Comment by alexr00** [comment]
@EhabY the way for you to solve this from your end would be to keep track of if your tree is currently refreshing and then wait to do the reveal until after the refreshing has finished.

**Comment by EhabY** [comment]
@alexr00 How is it possible to track if the tree is refreshing? IIRC the refresh events do not return a promise so I would need to keep track of all "open" nodes and then make sure that all of those open nodes have been reloaded (making sure to handle deleted nodes). Is there a better approach for this?

But even then, this would reduce the possibility of it happening instead of eliminating it which requires internal syncing.

**Comment by alexr00** [comment]
> How is it possible to track if the tree is refreshing?

I was thinking you could track if you have an unfinished `getChildren` method that was called.

**Comment by EhabY** [comment]
Ah, I actually tried that but it didn't work because refreshing would call `getChildren` in a recursive manner:

1. getChildren(undefined)
2. getChildren(root)
3. getChildren(child1)
...

So while we can easily check if a `getChildren` is being executed, we cannot know for sure when it has finished executing all calls unless we manually track all visible tree items (which can be error prone..)

**Comment by alexr00** [comment]
> So while we can easily check if a getChildren is being executed, we cannot know for sure when it has finished executing all calls unless we manually track all visible tree items (which can be error prone..)

I think this is still possible. You just need to increment a counter when `getChildren` is called, then decrement it before returning.

**Comment by EhabY** [comment]
I just attempted a solution as you described and it still fails, I have added a counter that is incremented first thing in `getChildren` and a decremented last thing in `getChildren` and I get the following trace. 

"Before" is printed first thing in `getChildren`, "During" is in the middle,"After" is printed last thing, "Done!" is printed when the counter is 0 and the created promise when refreshing is resolved.

```
Before:  0
During:  1
After:  0
Done! -- Finishes after only refreshing the root and not the rest of the tree
Before:  0
Before:  1
During:  2
After:  1
Before:  1
During:  2
After:  1
During:  1
After:  0
Done!
```

Notice how when I execute the command above, it refreshed the top level first (and finishes execution) before refreshing the rest of the tree - thus depending on the counter won't work as it finishes a refresh of the root before refreshing the rest.

**Comment by EhabY** [comment]
Are there any updates on this? The example I provided still fails consistently and in our case it makes the usage of `reveal` very problematic...

**Comment by eamodio** [comment]
I opened https://github.com/microsoft/vscode/pull/250239 to address this

**Comment by saltukkos** [comment]
@alexr00 the proposed workaround will not work. The issue is that in code
```
		const onDidChangeData = Event.debounce<TreeData<T>, { message: boolean; elements: (T | Root)[] }>(this._onDidChangeData.event, (result, current) => {
                // implementation ...
		}, 200, true);
```
there is a `debounce`, that effectively prevents the client code from understanding when `onDidChangeTreeData` will be handled.
And the root issue is that in case there is an active `reveal`, concurrent `onDidChangeData` will lead to a concurrent children fetching:

```
	private async fetchChildrenNodes(parentElement?: T): Promise<TreeNode[] | undefined> {
		// clear children cache
		this.clearChildren(parentElement);

                // get treeItems, executes concurrently
                // ...

		// register the elements, throws error for a second request 
		const items = treeItems.map((item, index) => item ? this.createAndRegisterTreeNode(coalescedElements[index], item, parentNode) : null);

		return coalesce(items);
      	}
```
And since client code can't reliably control the `onDidChangeTreeData` handling, it can happen at any time. I would suggest performing additional `this.clearChildren` right before registering items in cache (though I can't be sure if it's a good solution). Should we expect a fix? A lot of extensions are suffering from this issue.

The only possible workaround I can think of is to replicate `debounce` logic with >=200 ms timeout in a client code before firing the `onDidChangeTreeData`. This way, client can control the `onDidChangeData` handling and avoid internal debouncing. Then, client will be able to suppress `onDidChangeTreeData` events if there is an active `reveal`, but it's incredibly dirty hack that relies on internal vsCode's timeouts.

**Comment by saltukkos** [comment]
Another issue I found is happening when `getChildren` returns a node with `collapsibleState = TreeItemCollapsibleState.Expanded`. Then, if `reveal` happens before the next `getChildren` for this expanded node is finished, operation fails.
